### PR TITLE
Simplify how RunTests handles TFMs

### PR DESCRIPTION
**Pull request by jaredpar**
Previously our build scripts were manually passing TFMS into RunTests to select assemblies. That was incredibly fragile, particularly when we were updating the TFM set in the repo. It was really easy to miss tests this way.

Change the code to instead just say if we're requsting .NET Core, .NET Framework or both tests. That is a much more durable ask.

### CODE REVIEW

**Comment by jaredpar** [comment]
@RikkiGibson, @dibarbet  PTAL

File: src/Tools/Source/RunTests/Program.cs
Comment (jaredpar) [reply src/Tools/Source/RunTests/Program.cs:378]: No. I went back and forth about whether to hard code `net472` here. Think I will go back to that cause we hard code it everywhere else.

File: src/Tools/Source/RunTests/Program.cs
Comment (jasonmalinowski) [inline src/Tools/Source/RunTests/Program.cs:378]:
```suggestion
                    TestTargetFramework.Framework => dirName is "net4\d\d?",
```

I'm not sure if we'd ever have a reason to upgrade to net48 but if we do then this won't break that.

**Review by jasonmalinowski — APPROVED** [review]

File: src/Tools/Source/RunTests/Program.cs
Comment (jasonmalinowski) [reply src/Tools/Source/RunTests/Program.cs:378]: I see now you already said this to @dibarbet too, but it was collapsed. :smile:

File: src/Tools/Source/RunTests/Program.cs
Comment (jaredpar) [reply src/Tools/Source/RunTests/Program.cs:17]: Nope. Not sure how that snuck in there.

File: src/Tools/Source/RunTests/Program.cs
Comment (jaredpar) [reply src/Tools/Source/RunTests/Program.cs:379]: Not easily. This project is independent (doesn't ref other projects in the system) so it doesn't have access to that.

File: eng/build.sh
Comment (jaredpar) [inline eng/build.sh:340]: So not only was this change fragile, it was broken. Compiler hasn't been running our unit tests that targeted `net8.0` on Unix. And yes there are failures 😢

Comment (jaredpar) [reply eng/build.sh:340]: Will use https://github.com/dotnet/roslyn/issues/71571 to track the failing tests

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:307]: @tmat FYI I need to temporarily disable this assert that you added in November 

https://github.com/dotnet/roslyn/commit/8920679b0517fec747e6d165980f611306cc6444#diff-de1a7398366a42226adb835505958a0e044f9b9816301dcfb013f09820d91cd9

This is blocking a number of compiler tests on Unix (only for some reason). This PR is needed before I can merge the move to `net8.0` which is why I'm going for temp disable vs. dig into this right now.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:307]: Thanks. I keep meaning to file a bug on the fact that this is failing locally, only to get distracted by something else.

**Review by RikkiGibson — APPROVED** [review]

File: src/Tools/Source/RunTests/Options.cs
Comment (333fred) [inline src/Tools/Source/RunTests/Options.cs:171]: I'm not a _huge_ fan of naming this tfm, since it immediately reads to me that valid TFM strings shoudl be valid arguments, but I'm not going to block on it.

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs
Comment (tmat) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:307]: Ok.

File: src/Tools/Source/RunTests/Options.cs
Comment (333fred) [reply src/Tools/Source/RunTests/Options.cs:171]: I was thinking `runtime`

Comment (333fred) [reply src/Tools/Source/RunTests/Options.cs:171]: Or `testRuntime`

Comment (jaredpar) [reply src/Tools/Source/RunTests/Options.cs:171]: I don't have any objections to those names. This PR though is blocking several other infra PRs. Plan is to merge this as is assuming it's green. If it's red I'll incorporate the name suggestion on the next push, if it's green I'll try and incorporate it in one of the other infra PRS stacked behind it.

Comment (333fred) [reply src/Tools/Source/RunTests/Options.cs:171]: Understood. That's why I didn't block.

File: src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs:213]: `ExecutionConditionUtil.IsWindows`?

**Review by jjonescz — APPROVED** [review]

**Pull request merged**
